### PR TITLE
Implemented approval (partially) system

### PR DESCRIPTION
**Pull request by AshenI99**
Below issue came up while using `nft_is_approved` method inside `nft_transfer` function.
[Problem with @nearBindgen decorator](https://github.com/curaOS/contracts/issues/65)

So, had to use the internal function method to get rid of that error.

### CODE REVIEW

**Comment by AshenI99** [comment]
Below issue came up while using `nft_is_approved` method inside `nft_transfer` function.
[Problem with @nearBindgen decorator](https://github.com/curaOS/contracts/issues/65)

So, had to use the internal function method to get rid of that error.

File: as/cNFT/assembly/core.ts [hunk 1]
```diff
@@ -133,4 +137,81 @@ export function burn_design(token_id: TokenId): void {
...
+    token_id: TokenId,
+    approved_account_id: AccountId,
+    approval_id: string | null = null
+): boolean {
+
+    assert_token_exists(token_id);
+
+    let token = persistent_tokens.get(token_id)
+
+    assert(token.owner_id != approved_account_id, "Owner does not need any approval to perform transactions for the token")
```
Comment (achrafismyname) [inline as/cNFT/assembly/core.ts:205]: Is there a reason to throw an error here?
We can just return `true`

File: as/cNFT/assembly/core.ts [hunk 2]
```diff
@@ -133,4 +137,81 @@ export function burn_design(token_id: TokenId): void {
...
+    approved_account_id: AccountId,
+    approval_id: string | null = null
+): boolean {
+
+    return internal_nft_is_approved(
+            token_id,
+            approved_account_id,
+            approval_id
+        )
+}
```
Comment (achrafismyname) [inline as/cNFT/assembly/core.ts:173]: Why have two `nft_is_approved` functions, if they are going to the same thing anyway?

**Review by achrafismyname — COMMENTED** [review]
Seems good

File: as/cNFT/assembly/core.ts (hunk 1 above)
Comment (devey) [reply as/cNFT/assembly/core.ts:205]: True, you don't need this here being that is not part `nft_is_approved` logic and you are already checking that `token.owner_id != context.predecessor` in the `nft_transfer` method before calling this one.

File: as/cNFT/assembly/core.ts [hunk 3]
```diff
@@ -133,4 +137,81 @@ export function burn_design(token_id: TokenId): void {
...
+ * console.log(is_approved); // true | false
+ * ```
+ *
+ * @param token_id ID of the token
+ * @param approved_account_id ID of the account that need check whether it is approved or not for the given token
+ * @param approval_id Approval ID number for the given `approved_account_id`
+ * @return Whether the account is approved or not
+ */
+@nearBindgen
+export function nft_is_approved(
```
Comment (devey) [inline as/cNFT/assembly/core.ts:162]: `nft_is_approved` and it's internal are not part of _Core Standard_. You can move these two functions under a new file called `approval.ts`.

**Review by devey — CHANGES_REQUESTED** [review]
Looks good. A couple of adjustments and we can merge it.
Also make sure to add a test to _sandbox workspace_ after this and the workspace PR is closed to make sure tests are still working and `nft_is_approved` works as external call too (besides being used internally).

File: as/cNFT/assembly/core.ts (hunk 2 above)
Comment (AshenI99) [reply as/cNFT/assembly/core.ts:173]: Because of this error. https://github.com/curaOS/contracts/issues/65

Comment (AshenI99) [reply as/cNFT/assembly/core.ts:173]: Because of this error. https://github.com/curaOS/contracts/issues/65

**Review by devey — APPROVED** [review]

**Pull request merged**